sklearn/tree/_criterion.pyx

```diff
@@ -1283,9 +1290,12 @@ cdef class MAE(RegressionCriterion):
 
                 y_ik = y[i * self.y_stride + k]
 
-                impurity_left[0] += <double>fabs((<double> y_ik) -
-                                                 <double> median)
-        impurity_left[0] /= <double>((self.weighted_n_left) * self.n_outputs)
+                if sample_weight != NULL:
+                    w = sample_weight[i]
+
+                imp_left += fabs(y_ik - median) * w
```

Comment:
impurity_left and impurity_right point to array of type **double**.  It is very expensive to cast within the loop (implicitly) from DOUBLE_t to double on each iteration hence the introduction of holders imp_left and imp_right.  Cast is done at the end, outside of the loop where it is much less expensive and also will not suffer from loss in precision.

Reply:
DOUBLE_t is np.npy_float64 which is actually a double. So there is no casting cost because they are the same type. I am not really sure why we have all those casting around.

The only casting which we can make explicit is at the front of n_output which is an integer.

Reply:
It is confusing why DOUBLE_t is used throughout instead of double in that case?  Maybe some legacy reason (Python2? etc..?)

I'm not sure how intelligent the compiler is in determining that DOUBLE_t is np.npy_float64 and thus an implicit conversion is not required... so I ran some benchmarks to see the difference between updating directly to impurity_left/right[0]  versus imp_left/right the results are interesting (note I removed all **explicit** casts).  I ran the timings 5 times for each:

**imp_left/right**:
Seconds elapsed:|**22.688602447509766**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.552382230758667**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.5827374458313**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.657567262649536**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.656784534454346**
Total MAE: 0.8212764822134341

**impurity_left/right[0]** 
Seconds elapsed:|**25.91646981239319**
Total MAE: 0.8212764822134341

Seconds elapsed:|**26.169488668441772**
Total MAE: 0.8212764822134341

Seconds elapsed:|**25.957823753356934**
Total MAE: 0.8212764822134341

Seconds elapsed:|**26.088392734527588**
Total MAE: 0.8212764822134341

Seconds elapsed:|**25.923059225082397**
Total MAE: 0.8212764822134341

There is a notable difference.  This may not actually be caused by the implicit casts (if any?), it could the lookup to array item 0 for arrays impurity_left[0] / impurity_right[0] each iteration?

I think I can test this by changing my imp_left/right solution to have the impurity variable declared as a double instead of DOUBLE_t?  I'll try that and see what timings we get...

Reply:
Here are the results:
Seconds elapsed:|**22.72613835334778**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.563627243041992**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.58454728126526**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.605653047561646**
Total MAE: 0.8212764822134341

Seconds elapsed:|**22.564542770385742**
Total MAE: 0.8212764822134341

From this, I think we can safely conclude that there are no issues with implicit casts, the time wasted is looking up item 0 in the arrays on each iteration.  Therefore I propose we keep imp_left / imp_right.

Reply:
_(As an aside, I decided to run once last benchmark, explicitly casting self.n_outputs to type double and got the following:)
Seconds elapsed:|22.798115253448486
Seconds elapsed:|22.661853551864624
Seconds elapsed:|22.832958459854126
Seconds elapsed:|22.838365077972412
Seconds elapsed:|22.705886602401733
(so not overly conclusive and thus no benefit in introducing an explicit cast here)_

Reply:
DOUBLE_t and double are the same at the end of the day. However, you point regarding the dereferencing is actually interesting and then it makes sense to avoid doing this in the loop and just make a single affectation at the end

Reply:
If DOUBLE_t and double are exactly the same, we should replace all DOUBLE_t with double, and the equivalent with SIZE_t if equivalent to another native datatype?  Otherwise it will just cause further confusion (as above) to other developers and untrue assumptions.  Unless anyone knows of some other good reason for these?

Reply:
Basically, in C++ double_t is an alias to double as well: http://www.cplusplus.com/reference/cmath/double_t/.

So I don't think that it worth making do much change. I don't exactly the reason why the code was originally written that way but it could have been imposed with Cython in the past.